streams/shared/src/main/scala/zio/stream/ZSink.scala

```diff
@@ -578,108 +579,106 @@ object ZSink extends ZSinkPlatformSpecificConstructors {
         } yield push
       }
     else
-      ZSink(ZManaged.succeed(_ => Push.emit(z)))
+      ZSink(ZManaged.succeed(_ => Push.emit(z, Chunk.empty)))
 
   /**
    * A sink that effectfully folds its inputs with the provided function, termination predicate and initial state.
    *
    * This sink may terminate in the middle of a chunk and discard the rest of it. See the discussion on the
    * ZSink class scaladoc on sinks vs. transducers.
    */
-  def foldM[R, E, I, S](z: S)(contFn: S => Boolean)(f: (S, I) => ZIO[R, E, S]): ZSink[R, E, I, S] =
+  //TODO:  implementation is wrong now
+  def foldM[R, E, I, S](z: S)(contFn: S => Boolean)(f: (S, I) => ZIO[R, E, S]): ZSink[R, E, I, Nothing, S] =
     foldChunksM(z)(contFn)((s, is) => is.foldWhileM(s)(contFn)(f))
 
   /**
    * A sink that folds its inputs with the provided function and initial state.
    */
-  def foldLeft[I, S](z: S)(f: (S, I) => S): ZSink[Any, Nothing, I, S] =
+  def foldLeft[I, S](z: S)(f: (S, I) => S): ZSink[Any, Nothing, I, Nothing, S] =
     fold(z)(_ => true)(f)
 
   /**
    * A sink that folds its input chunks with the provided function and initial state.
    */
-  def foldLeftChunks[I, S](z: S)(f: (S, Chunk[I]) => S): ZSink[Any, Nothing, I, S] =
+  def foldLeftChunks[I, S](z: S)(f: (S, Chunk[I]) => S): ZSink[Any, Nothing, I, Nothing, S] =
     foldChunks(z)(_ => true)(f)
 
   /**
    * A sink that effectfully folds its input chunks with the provided function and initial state.
    */
-  def foldLeftChunksM[R, E, I, S](z: S)(f: (S, Chunk[I]) => ZIO[R, E, S]): ZSink[R, E, I, S] =
+  def foldLeftChunksM[R, E, I, S](z: S)(f: (S, Chunk[I]) => ZIO[R, E, S]): ZSink[R, E, I, Nothing, S] =
     foldChunksM[R, E, I, S](z: S)(_ => true)(f)
 
   /**
    * A sink that effectfully folds its inputs with the provided function and initial state.
    */
-  def foldLeftM[R, E, I, S](z: S)(f: (S, I) => ZIO[R, E, S]): ZSink[R, E, I, S] =
+  def foldLeftM[R, E, I, S](z: S)(f: (S, I) => ZIO[R, E, S]): ZSink[R, E, I, Nothing, S] =
     foldM[R, E, I, S](z: S)(_ => true)(f)
 
   /**
    * A sink that executes the provided effectful function for every element fed to it.
    */
-  def foreach[R, E, I](f: I => ZIO[R, E, Any]): ZSink[R, E, I, Unit] =
-    ZSink.fromPush {
-      case Some(is) => is.mapM_(f).mapError(Left(_))
-      case None     => ZIO.fail(Right(()))
+  def foreach[R, E, I](f: I => ZIO[R, E, Any]): ZSink[R, E, I, Nothing, Unit] =
+    ZSink.fromPush[R, E, I, Nothing, Unit] {
+      case Some(is) => is.mapM_(f).mapError(e => (Left(e), Chunk.empty))
+      case None     => Push.emit((), Chunk.empty)
     }
 
   /**
    * Creates a single-value sink produced from an effect
    */
-  def fromEffect[R, E, Z](b: => ZIO[R, E, Z]): ZSink[R, E, Any, Z] =
-    fromPush(_ => b.foldM(Push.fail, Push.emit))
+  def fromEffect[R, E, Z](b: => ZIO[R, E, Z]): ZSink[R, E, Any, Nothing, Z] =
+    fromPush[R, E, Any, Nothing, Z](_ => b.foldM(Push.fail(_, Chunk.empty), z => Push.emit(z, Chunk.empty)))
```

Comment:
This (and ZSink.die, ZSink.succeed, and basically all the constant-valued sinks) loses leftovers.